File: Cpp/Min Stack.cpp
// 31ms, 30.08%
class MinStack
{
private:
    vector<int> s1;
    vector<int> s2;
    
public:
    MinStack()
    {
    }
    
    void push(int x)
    {
        s1.push_back(x);
        if(s2.empty() || x <= s2.back()) s2.push_back(x);
    }
    
    void pop()
    {
        if(s2.back() == s1.back()) s2.pop_back();
        s1.pop_back();
    }
    
    int top()
    {
        return s1.back();
    }
    
    int getMin()
    {
        return s2.back();
    }
};

File: Python3/Min Stack.py
# 60ms, 100.0%
class MinStack:

    def __init__(self):
        self.min_stack = list()    # 'self.min_stack = self.min_record = []' is NOT correct
        self.min_record = list()   # that will make them two names of the same list
        
    def push(self, x):
        self.min_stack.append(x)
        if not self.min_record or x <= self.min_record[-1]: self.min_record.append(x)
        
    def pop(self):
        popped = self.min_stack.pop()
        if popped == self.min_record[-1]: self.min_record.pop()
        
    def top(self):
        return self.min_stack[-1]
        
    def getMin(self):
        return self.min_record[-1]
        
